plugin/vcs/gitlab/gitlab.go

```diff
@@ -182,6 +190,100 @@ func (provider *Provider) APIURL(instanceURL string) string {
 	return fmt.Sprintf("%s/%s", instanceURL, apiPath)
 }
 
+// ExchangeOAuthToken exchange oauth content with the provdied authentication code
+func (provider *Provider) ExchangeOAuthToken(ctx context.Context, instanceURL string, oauthExchange common.OAuthExchange, code string, redirectURL string) (*common.OAuthToken, error) {
+	urlParams := &url.Values{}
+	urlParams.Set("client_id", oauthExchange.ClientID)
+	urlParams.Set("client_secret", oauthExchange.ClientSecret)
+	urlParams.Set("code", code)
+	urlParams.Set("redirect_uri", redirectURL)
+	urlParams.Set("grant_type", "authorization_code")
+	url := fmt.Sprintf("%s/oauth/token?%s", instanceURL, urlParams.Encode())
+
+	req, err := http.NewRequest("POST", url, nil)
+
+	if err != nil {
+		urlParams.Set("client_secrete", "**encrypted**")
+		urlWithoutSecret := fmt.Sprintf("%s/oauth/token?%s", instanceURL, urlParams.Encode())
+		return nil, fmt.Errorf("failed to construct POST %v (%w)", urlWithoutSecret, err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to exchange Oauth Token, code %v, error: %v", resp.StatusCode, err)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to read oauth response body, code %v, error: %v", resp.StatusCode, err)
+	}
+
+	oauthToken := &common.OAuthToken{}
+	if err := json.Unmarshal(body, oauthToken); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal oauth response body, code %v, error: %v", resp.StatusCode, err)
+	}
+
+	// derivative expiresAt
+	// For GitLab, as of 13.12, the default config won't expire the access token, thus this field is 0.
+	// see https://gitlab.com/gitlab-org/gitlab/-/issues/21745.
+	if oauthToken.ExpiresIn != 0 {
+		oauthToken.ExpiresTs = oauthToken.CreatedAt + oauthToken.ExpiresIn
+	}
+
+	return oauthToken, nil
+}
+
+// FetchRepositoryList will fetch all repository within a given user's scope
```

Comment:
For GitLab, we should leave a comment here that says we only return the project in which the token owner is at least the maintainer.